### PR TITLE
Added `std/string.cup` string manipulation library

### PR DESCRIPTION
**Pull request by Weltspear**
`str` struct has support for string slicing and has a lot of useful functions.

### CODE REVIEW

File: std/string.cup [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+    
+    return self;
+}
+
+// Implicit destructor
+// You should call this before freeing str itself, if you don't do this you will get a memory leak.
+method str::drop(){
+    free(self.chars);
+    self.isdropped = true;
+}
```
Comment (mustafaquraish) [inline std/string.cup]: Maybe we can have a helper called `str::free()` that both drops the chars and frees the string itself. Is there a use case where we would want to drop but not free? If not, we can just make this one free the `str` object too by calling `free(self)`

File: std/string.cup [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+method str::contains_cstr(other: char*): i32{
+    let temp_str: str* = new str(other);
+    let result = self::contains(temp_str);
+    temp_str::drop();
+    free(temp_str);
+    if (result == -1){
+        return false;
+    }
+    else 
+        return true;
```
Comment (mustafaquraish) [inline std/string.cup]: this can just be `return result != -1`

File: std/string.cup [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+    let result = self::contains_idx(other);
+    
+    if (result == -1){
+        return false;
+    }
+    else 
+        return true;
+}
+
+method str::drop();
```
Comment (mustafaquraish) [inline std/string.cup]: Can just move the method definition up here since it doesn't need to use any of the other methods.

File: std/string.cup [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+}
+
+method str::contains(other: str*): i32{
+    let result = self::contains_idx(other);
+    
+    if (result == -1){
+        return false;
+    }
+    else 
+        return true;
```
Comment (mustafaquraish) [inline std/string.cup]: `return self::contains_idx(other) != -1`

File: std/string.cup [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+}
+
+method str::startswith(prefix: str*): i32{
+    return self::startswith_cstr(prefix.chars);
+}
+
+method str::equals_cstr(other: char*): i32{
+    return streq(self.chars, other);
+}
+
```
Comment (mustafaquraish) [inline std/string.cup:157]: This feels weird, Is the `str` type null terminated? I would expect it to not, if it's explicitly storing length.

File: std/string.cup [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+    for (let i = first; i < self.len; ++i){
+        self.chars[i] = sl::get(i-first);
+    }
+    
+    self.chars[self.len] = '\0';
+}
+
+method str::len(): i32{
+    return self.len;
+}
```
Comment (mustafaquraish) [inline std/string.cup]: Is there any point to having this helper as opposed to just doing `str.len` where you want it?

**Comment by mustafaquraish** [comment]
A lot of cool stuff! Left some comments, but it would be nice to add some tests as well :)

**Comment by Weltspear** [comment]
Okay

File: std/string.cup (hunk 5 above)
Comment (Weltspear) [reply std/string.cup:157]: Yeah it has null terminator.

Comment (Weltspear) [reply std/string.cup:157]: Len is length of string without counting the null terminator

File: std/string.cup (hunk 6 above)
Comment (Weltspear) [reply std/string.cup]: It looks nicer.

File: std/string.cup (hunk 1 above)
Comment (Weltspear) [reply std/string.cup]: Okay

File: std/string.cup (hunk 3 above)
Comment (Weltspear) [reply std/string.cup]: Okay

File: std/string.cup (hunk 4 above)
Comment (Weltspear) [reply std/string.cup]: Okay

File: std/string.cup (hunk 5 above)
Comment (Weltspear) [reply std/string.cup:157]: Essentially you can say that `str` is a wrapper of `char*` and it's `self.chars` are compatible with `cstr` functions.

Comment (mustafaquraish) [reply std/string.cup:157]: Fair enough

Comment (Weltspear) [reply std/string.cup:157]: It keeps the `len` to not to call `strlen` everytime you need to get string length.

File: std/string.cup (hunk 2 above)
Comment (Weltspear) [reply std/string.cup]: Okay

**Comment by mustafaquraish** [comment]
The new changes look good to me, will merge once you've added some tests!

**Comment by Weltspear** [comment]
Yeah I know I'll add them when I can.

**Comment by Weltspear** [comment]
Can I create a scripts for running the compiler on termux? Cause I managed to make it run there somehow.

**Comment by Weltspear** [comment]
Or nvm there are some issues probably with command execution.

**Comment by Weltspear** [comment]
For some reason it says permission denied

**Comment by mustafaquraish** [comment]
> For some reason it says permission denied

You need to mark the file as executable. `chmod +x tests/string.sh`

**Comment by Weltspear** [comment]
Huh on my machine it worked without chmoding.

**Comment by Weltspear** [comment]
Also how do I commit the chmoded file cause git doesn't detect this change?

**Pull request merged**